### PR TITLE
common-cli: arguments parsing improvement

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: common-cli/src/test/java/co/cask/common/cli/CommandMatchTest.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+    @Override
+    public String getDescription() {
+      return "Creates the cluster";
+    }
+  };
+
+  @Test
+  public void getArgumentsAllOptionalTest() throws NoSuchFieldException, IllegalAccessException {
+    String testInput = new StringBuilder("create cluster ")
```
Comment (jwang47) [inline common-cli/src/test/java/co/cask/common/cli/CommandMatchTest.java]: Use `String.format("create cluster %s with template %s etc")`. It's easier to read.

File: common-cli/src/test/java/co/cask/common/cli/CommandSetTest.java [hunk 2]
```diff
@@ -110,13 +110,14 @@ public String getDescription() {
     };
 
     CommandSet commandSet = new CommandSet<Command>(ImmutableList.of(greetCommand));
```
Comment (jwang47) [inline common-cli/src/test/java/co/cask/common/cli/CommandSetTest.java:112]: Probably also want to add a test case here for testing what input matches which command patterns. I think I originally did the matching based on the command pattern prefix before the first argument (e.g. "greet " for "greet <user>"), and I don't think I saw any changes to that.

For example, test the input "greet bob times 3" against two commands with the patterns "greet <user> times <times>" and "greet <user> timestamp <timestamp>". Probably want to add even more test cases, but that's a start.

File: common-cli/src/main/java/co/cask/common/cli/CommandMatch.java [hunk 3]
```diff
@@ -46,32 +54,115 @@ public Command getCommand() {
...
   }
 
-  private Arguments parseArguments(String input, String pattern) {
-    String[] patternTokens = pattern.split(" ");
-    String[] inputTokens = input.split(" ");
+  public String getInput() {
+    return input;
+  }
 
+  private Arguments parseArguments(String input, String pattern) {
```
Comment (chtyim) [inline common-cli/src/main/java/co/cask/common/cli/CommandMatch.java:73]: Please add more comment to describe what's the logic is. The code in this method is not obvious what it is doing.

File: common-cli/src/main/java/co/cask/common/cli/CommandMatch.java [hunk 4]
```diff
@@ -46,32 +54,115 @@ public Command getCommand() {
...
-          args.put(patternToken.substring(1, patternToken.length() - 1), inputTokens[i]);
+    if (optionalPartFirstArgStart != 0) {
+      return input.substring(currentIndex, input.indexOf(
+        optionalPatternPart.substring(0, optionalPartFirstArgStart), currentIndex));
+    } else {
+      return input.substring(currentIndex);
+    }
+  }
+
+  private int parseArgs(ImmutableMap.Builder<String, String> args, String input,
```
Comment (chtyim) [inline common-cli/src/main/java/co/cask/common/cli/CommandMatch.java]: Same here. Please have comment on what the logic is.

File: common-cli/src/test/java/co/cask/common/cli/CommandMatchTest.java (hunk 1 above)
Comment (ghost) [inline common-cli/src/test/java/co/cask/common/cli/CommandMatchTest.java]: ok

File: common-cli/src/test/java/co/cask/common/cli/CommandSetTest.java (hunk 2 above)
Comment (ghost) [inline common-cli/src/test/java/co/cask/common/cli/CommandSetTest.java:112]: Also added this test in https://github.com/caskdata/common/pull/6

**Comment by ghost** [comment]
Changed parse logic. Made parse logic more clear for understanding.

File: common-cli/src/main/java/co/cask/common/cli/CommandMatch.java [hunk 5]
```diff
@@ -46,32 +56,201 @@ public Command getCommand() {
...
-      if (patternToken.startsWith("<") && patternToken.endsWith(">")) {
-        // required argument
-        args.put(patternToken.substring(1, patternToken.length() - 1), inputTokens[i]);
-      } else if (patternToken.startsWith("[") && patternToken.endsWith("]")) {
-        // optional argument
-        if (i < inputTokens.length) {
-          args.put(patternToken.substring(1, patternToken.length() - 1), inputTokens[i]);
-        }
+    List<String> splitInput = Parser.parseInput(input);
+    List<String> splitPattern = Parser.parsePattern(pattern);
```
Comment (jwang47) [inline common-cli/src/main/java/co/cask/common/cli/CommandMatch.java:77]: Can you add some more detailed javadocs explaining what `Parser.parseInput` and `Parser.parsePattern` do? I see there is some, but those don't explain what the output actually is.

**Comment by jwang47** [comment]
We will also have to change the completer logic in `PrefixCompleter` and how the `PrefixCompleter`s are generated in `CLI.generateCompleters` to account for this change.

Currently, we're assuming that each argument is only a single word and also we're not accounting for multiple optional arguments in a row.

File: common-cli/src/main/java/co/cask/common/cli/CommandMatch.java (hunk 5 above)
Comment (ghost) [inline common-cli/src/main/java/co/cask/common/cli/CommandMatch.java:77]: Added

**Comment by ghost** [comment]
Improved java-docs.
I think, it is better to create new PR for changing Completers logic.

**Comment by jwang47** [comment]
LGTM though seems like we may be able to simplify some of the code later.

**Pull request merged**